src/main/java/com/auth0/TransientCookieStore.java

```diff
@@ -0,0 +1,154 @@
+        store(response, StorageUtils.STATE_KEY, state, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Stores a nonce value as a cookie on the response.
+     *
+     * @param response the response object to set the cookie on
+     * @param nonce the value for the nonce cookie. If null, no cookie will be set.
+     * @param sameSite the value for the SameSite attribute on the cookie
+     * @param useLegacySameSiteCookie whether to set a fallback cookie or not
+     */
+    static void storeNonce(HttpServletResponse response, String nonce, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        store(response, StorageUtils.NONCE_KEY, nonce, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the state cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the state cookie, if it exists
+     */
+    static String getState(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.STATE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the nonce cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the nonce cookie, if it exists
+     */
+    static String getNonce(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.NONCE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    private static void store(HttpServletResponse response, String key, String value, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        Validate.notNull(response, "response must not be null");
+        Validate.notNull(key, "key must not be null");
+        Validate.notNull(sameSite, "sameSite must not be null");
+
+        if (value == null) {
+            return;
+        }
+
+        boolean sameSiteNone = SameSite.NONE == sameSite;
+
+        String cookie = String.format("%s=%s; HttpOnly; Max-Age=%d; SameSite=%s", key, value, MAX_AGE_SECONDS, sameSite.getValue());
+        if (sameSiteNone) {
+            cookie = cookie.concat("; Secure");
+        }
+
+        // Servlet Cookie API does not yet support setting the SameSite attribute, so just set cookie on header
+        response.addHeader("Set-Cookie", cookie);
+
+        // set legacy fallback cookie (if configured) for clients that won't accept SameSite=None
+        if (sameSiteNone && useLegacySameSiteCookie) {
+            String legacyCookie = String.format("%s=%s; HttpOnly; Max-Age=%d", "_" + key, value, MAX_AGE_SECONDS);
```

Comment:
**Nit:** I would suggest building out the base value - `"%s=%s; HttpOnly; Max-Age=%d"` - once and then append to that for the SameSite/Secure additions. A little more DRY.

Reply:
Good suggestion, will update.

Reply:
Actually it might be even better to represent the cookie string as a new (internal) class that can build its header string representation. So could be used like:

```java
AuthCookie sameSiteCookie = new AuthCookie(key, value);
sameSiteCookie.setSameSite(sameSite.getValue());
sameSiteCookie.setSecure(isSameSiteNone);

// Servlet Cookie API does not yet support setting the SameSite attribute, so just set cookie on header
response.addHeader("Set-Cookie", sameSiteCookie.buildHeaderString());

// set legacy fallback cookie (if configured) for clients that won't accept SameSite=None
if (isSameSiteNone && useLegacySameSiteCookie) {
    AuthCookie legacyCookie = new AuthCookie("_" + key, value);
    response.addHeader("Set-Cookie", legacyCookie.buildHeaderString());
}
```

Reply:
Sounds like the right solution for Java, another class 😆 but yes, I like that, more consistent.

Reply:
Updated with a new `AuthCookie` class.